File: CPP/ByRound/2038/2038_G_Guess_One_Character.cpp
// Problem: CF 2038 G - Guess One Character
// https://codeforces.com/contest/2038/problem/G

/*
Purpose: This code solves the interactive problem "Guess One Character" where
the goal is to guess at least one character of a hidden binary string s of
length n using at most 3 queries.

Algorithms/Techniques:
- Use three specific queries to determine the pattern:
    1. Query "1" (count of occurrences of '1')
    2. Query "10" (count of occurrences of "10")
    3. Query "11" (count of occurrences of "11")
- Based on these counts, we can deduce information about the structure of the
string and determine the value of at least one character.
- The logic to infer the result is based on:
    - If the count of "11" is 0, then there are no consecutive '1's.
    - If the count of "10" is greater than or equal to the count of "11", then
the last character must be '0' if the last two characters are "10".
    - The final guess is made by analyzing how many '1's are present compared to
the count of "10" and "11".

Time Complexity: O(1) - Fixed number of operations regardless of input size.
Space Complexity: O(1) - Only a constant amount of space is used.
*/

#include <cstdio>
#include <iostream>

#define inf 0x3f3f3f3f
using namespace std;
int n, p;
void work() {
  cin >> n;
  // Query for count of "1", "10", and "11"
  cout << "1 1\n1 10\n1 11\n";
  fflush(stdout);
  int a, b, c;
  cin >> a >> b >> c; // Read the responses to queries
  // The logic to determine the value of the last character
  cout << "0 " << n << ' ' << (a != b + c) << '\n';
  fflush(stdout);
  cin >> p;
}
signed main() {
  int _ = 1;
  cin >> _;
  while (_--)
    work();
}

// https://github.com/VaHiX/CodeForces/

File: Python/ByTier/G/2038_G_Guess_One_Character.py
# Problem: CF 2038 G - Guess One Character
# https://codeforces.com/contest/2038/problem/G

"""
Algorithm: Guess One Character
Techniques: Interactive Binary Search, String Matching, Case Analysis

Time Complexity: O(1) per test case (at most 4 queries)
Space Complexity: O(1) (only using a few variables)

The solution uses a series of queries to determine the value of at least one character in the binary string.
It leverages the relationship between counts of substrings "01", "10", "0", "1", and "11" to deduce the string's composition.
"""

import sys

input = lambda: sys.stdin.readline().rstrip()


def ask(t):
    # Ask the jury about the count of substring t in s
    print(f"1 {t }", flush=True)
    r = int(input())
    if r == -1:
        exit()
    return r


def answer(i, c):
    # Submit the guess that s[i] = c
    print(f"0 {i } {c }", flush=True)
    r = int(input())
    if r == -1:
        exit()


def solve_case():
    _ = int(input())

    # Query count of "01" substring
    c01 = ask("01")
    if c01 == 0:
        # If "01" is never found, the string doesn't have "0" followed by "1"
        # So we check for "10" to determine the pattern
        c10 = ask("10")
        if c10 > 0:
            # If "10" is found, then string has a "1" followed by "0"
            # So first character must be "1"
            answer(1, "1")
        else:
            # If neither "01" nor "10" is found, all characters are same
            # Query count of "0"
            c0 = ask("0")
            if c0 > 0:
                # If "0" is found, then string is all zeros
                answer(1, "0")
            else:
                # If "0" is not found, then string is all ones
                answer(1, "1")
    else:
        # If "01" is found, we can infer something about the structure
        # Query count of "1"
        c1 = ask("1")
        # Query count of "11"
        c11 = ask("11")
        # Compare counts to deduce the first character
        if c11 == c1 - c01:
            # If "11" count equals "1" count minus "01" count, 
            # then the string starts with "0"
            answer(1, "0")
        else:
            # Otherwise, it starts with "1"
            answer(1, "1")


for _ in range(int(input())):
    solve_case()


# https://github.com/VaHiX/CodeForces/